platform/mbed_retarget.cpp

```diff
@@ -112,20 +118,138 @@ void remove_filehandle(FileHandle *file) {
 extern serial_t stdio_uart;
-#if MBED_CONF_PLATFORM_STDIO_CONVERT_NEWLINES
-static char stdio_in_prev;
-static char stdio_out_prev;
-#endif
 #endif
 
-static void init_serial() {
-#if DEVICE_SERIAL
+/* Private FileHandle to implement backwards-compatible functionality of
+ * direct HAL serial access for default stdin/stdout/stderr.
+ * This is not a particularly well-behaved FileHandle for a stream, which
+ * is why it's not public. People should be using UARTSerial.
+ */
+class DirectSerial : public FileHandle {
+public:
+    DirectSerial(PinName tx, PinName rx, int baud);
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual off_t seek(off_t offset, int whence = SEEK_SET) {
+        return -ESPIPE;
+    }
+    virtual off_t size() {
+        return -EINVAL;
+    }
+    virtual int isatty() {
+        return true;
+    }
+    virtual int close() {
+        return 0;
+    }
+    virtual short poll(short events) const;
+};
+
+DirectSerial::DirectSerial(PinName tx, PinName rx, int baud) {
     if (stdio_uart_inited) return;
-    serial_init(&stdio_uart, STDIO_UART_TX, STDIO_UART_RX);
-#if MBED_CONF_PLATFORM_STDIO_BAUD_RATE
-    serial_baud(&stdio_uart, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
-#endif
+    serial_init(&stdio_uart, tx, rx);
+    serial_baud(&stdio_uart, baud);
+}
+
+ssize_t DirectSerial::write(const void *buffer, size_t size) {
+    const unsigned char *buf = static_cast<const unsigned char *>(buffer);
+    for (size_t i = 0; i < size; i++) {
+        serial_putc(&stdio_uart, buf[i]);
+    }
+    return size;
+}
+
+ssize_t DirectSerial::read(void *buffer, size_t size) {
+    unsigned char *buf = static_cast<unsigned char *>(buffer);
+    if (size == 0) {
+        return 0;
+    }
+    buf[0] = serial_getc(&stdio_uart);
+    return 1;
+}
+
+short DirectSerial::poll(short events) const {
+    short revents = 0;
+    if ((events & POLLIN) && serial_readable(&stdio_uart)) {
+        revents |= POLLIN;
+    }
+    if ((events & POLLOUT) && serial_writable(&stdio_uart)) {
+        revents |= POLLOUT;
+    }
+    return revents;
+}
+
+class Sink : public FileHandle {
+public:
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual off_t seek(off_t offset, int whence = SEEK_SET) { return ESPIPE; }
+    virtual off_t size() { return -EINVAL; }
+    virtual int isatty() { return true; }
+    virtual int close() { return 0; }
+};
+
+ssize_t Sink::write(const void *buffer, size_t size) {
+    // Just swallow the data - this is historical non-DEVICE_SERIAL behaviour
+    return size;
+}
+
+ssize_t Sink::read(void *buffer, size_t size) {
+    // Produce 1 zero byte - historical behaviour returned 1 without touching
+    // the buffer
+    unsigned char *buf = static_cast<unsigned char *>(buffer);
+    buf[0] = 0;
+    return 1;
+}
+
+
+MBED_WEAK FileHandle* mbed::mbed_target_override_console(int fd)
```

Comment:
nit: pointer on right:
``` diff
- MBED_WEAK FileHandle *mbed::mbed_target_override_console(int fd)
+ MBED_WEAK FileHandle *mbed::mbed_target_override_console(int fd)
```

Reply:
I agree, but I just changed it to left for Martin.

See his comment 4 days ago - there's a pending style guide change saying it should be that way for return types (but not parameters or variable declarations).

Reply:
That's how it has been (most of files have it that way - examples found in platform/) - I just captured it in the Handbook as was asked recently about the function/methods return types alignment for pointers/references.

Reply:
My feeling is older code like much of drivers and platform tends to be `Foo* func` - more recent code tends to be `Foo *func` because that's how the style guide has been for a few years. Certainly the code base is inconsistent, but I thought `*`s were gradually moving right. I've only ever put `*` or `&` on the left in mbed OS code if it's to match immediately-surrounding existing code.

Reply:
Lets take it here https://github.com/ARMmbed/Handbook/pull/382 and do a final approval (appreciate reviews there)

That is my view as well (older vs newer) therefore I proposed the style update (to keep it as it has been)

Reply:
Oh, sorry for the ping pong game @kjbracey-arm. Will move to https://github.com/ARMmbed/Handbook/pull/382.

We can just fix this with a script later.